numpy/f2py/src/fortranobject.c

```diff
@@ -720,8 +721,9 @@ PyArrayObject* array_from_pyobj(const int type_num,
             /* intent(cache) */
             if (PyArray_ISONESEGMENT(arr)
                 && PyArray_ITEMSIZE(arr)>=elsize) {
-                if (check_and_fix_dimensions(arr,rank,dims)) {
-                    return NULL; /*XXX: set exception */
+                if (check_and_fix_dimensions(arr, rank, dims, mess, sizeof(mess))) {
+                    PyErr_SetString(PyExc_ValueError, mess);
```

Comment:
I pushed an update that does this.  I kept the buffer as an argument of `check_and_fix_dimensions()`, because `array_from_pyobj()` already has the buffer `mess` allocated, and it uses it for other exception messages.